x-pack/solutions/security/plugins/security_solution/server/lib/siem_migrations/rules/api/util/installation.ts

```diff
@@ -169,51 +166,60 @@ export const installTranslated = async ({
   securitySolutionContext,
   rulesClient,
   savedObjectsClient,
-  logger,
-}: InstallTranslatedProps) => {
+}: InstallTranslatedProps): Promise<number> => {
   const detectionRulesClient = securitySolutionContext.getDetectionRulesClient();
   const ruleMigrationsClient = securitySolutionContext.getSiemRuleMigrationsClient();
 
-  const { data: rulesToInstall } = await ruleMigrationsClient.data.rules.get(migrationId, {
-    filters: { ids, installable: true },
-    from: 0,
-    size: MAX_TRANSLATED_RULES_TO_INSTALL,
-  });
+  let installedCount = 0;
+  const installationErrors: Error[] = [];
 
-  const { customRulesToInstall, prebuiltRulesToInstall } = rulesToInstall.reduce(
-    (acc, item) => {
-      if (item.elastic_rule?.prebuilt_rule_id) {
-        acc.prebuiltRulesToInstall.push(item);
-      } else {
-        acc.customRulesToInstall.push(item);
-      }
-      return acc;
-    },
-    { customRulesToInstall: [], prebuiltRulesToInstall: [] } as {
-      customRulesToInstall: StoredRuleMigration[];
-      prebuiltRulesToInstall: StoredRuleMigration[];
-    }
-  );
+  // Install rules that matched Elastic prebuilt rules
+  const prebuiltRuleBatches = ruleMigrationsClient.data.rules.searchBatches(migrationId, {
+    filters: { ids, installable: true, prebuilt: true },
+  });
+  let prebuiltRulesToInstall = await prebuiltRuleBatches.next();
+  while (prebuiltRulesToInstall.length) {
+    const { rulesToUpdate, errors } = await installPrebuiltRules(
+      prebuiltRulesToInstall,
+      enabled,
+      securitySolutionContext,
+      rulesClient,
+      savedObjectsClient,
+      detectionRulesClient
+    );
+    installedCount += rulesToUpdate.length;
+    installationErrors.push(...errors);
+    await ruleMigrationsClient.data.rules.update(rulesToUpdate);
+    prebuiltRulesToInstall = await prebuiltRuleBatches.next();
+  }
 
-  const updatedPrebuiltRules = await installPrebuiltRules(
-    prebuiltRulesToInstall,
-    enabled,
-    securitySolutionContext,
-    rulesClient,
-    savedObjectsClient,
-    detectionRulesClient
-  );
+  // Install rules with custom translation
+  const customRuleBatches = ruleMigrationsClient.data.rules.searchBatches(migrationId, {
+    filters: { ids, installable: true, prebuilt: false },
+  });
+  let customRulesToInstall = await customRuleBatches.next();
+  while (customRulesToInstall.length) {
+    const { rulesToUpdate, errors } = await installCustomRules(
+      customRulesToInstall,
+      enabled,
+      detectionRulesClient
+    );
+    installedCount += rulesToUpdate.length;
+    installationErrors.push(...errors);
+    await ruleMigrationsClient.data.rules.update(rulesToUpdate);
+    customRulesToInstall = await customRuleBatches.next();
+  }
 
-  const updatedCustomRules = await installCustomRules(
-    customRulesToInstall,
-    enabled,
-    detectionRulesClient,
-    logger
+  const { error: installTimelinesError } = await performTimelinesInstallation(
```

Comment:
Would it make sense to call this right after `installPrebuiltRules` and only if `installedCount > 0`

Reply:
Together with the throwing an exception in case of timelines installation error? If we will throw an exception there we will skip custom rules installation

Reply:
no, the throwing is okay at the end. I mean just the function call, to do it only if necessary (some prebuilt rule were installed).

Reply:
done in https://github.com/elastic/kibana/pull/206658/commits/be42922e1e6fc7dbd4b5dc63dbb3d81b9b2ad308